lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+          || (CurrentArgument.Offset + CurrentSize) % NextAlignment != 0) {
+        // Abandon any cases where natural alignment causes a padding "hole"
+        // to be required.
+        //
+        // TODO: we should probably preprocess the input to fill such holes in
+        //       before trying to convert the functions.
+        return std::nullopt;
+      }
+
+      // Create the argument from this field.
       model::Argument &New = Result.emplace_back();
       New.Index = IndexOffset++;
-      New.Type = Field.Type;
-      New.CustomName = Field.CustomName;
-      New.OriginalName = Field.OriginalName;
+      New.Type = CurrentArgument.Type;
+      New.CustomName = CurrentArgument.CustomName;
+      New.OriginalName = CurrentArgument.OriginalName;
+
+      CurrentAlignment = NextAlignment;
     }
 
     return Result;
   }
 
+  /// Helper used for converting return value representation to
+  /// the c-style
+  ///
+  /// \param Used a set of registers confirmed to be in use by
+  ///        the function in question.
+  /// \param Allowed an ordered list of registers allowed to be used
+  ///        for returning values by the ABI.
+  /// \param PointerToCopyLocation an optional register used for pointing to
+  ///        return values in memory on some architectures. It is set to
+  ///        `model::Register::Invalid` if it's not applicable.
+  /// \param Binary a const reference to the binary containing information
+  ///        about already present types.
+  /// \param TypeDependencies a list of types that need to be added to the model
+  ///        for the returned return value to be valid.
+  ///
+  /// \tparam RegisterType type of a register, should be deduced from
+  ///         the \ref Used.
+  /// \tparam RegisterCount the number of registers for the current ABI, should
+  ///         be auto-magically deduced from the \ref Allowed.
+  ///
+  /// \return
   template<typename RegisterType, size_t RegisterCount, bool DryRun = false>
   static std::optional<model::QualifiedType>
-  convertReturnValue(const SortedVector<RegisterType> &UsedRegisters,
-                     const RegisterArray<RegisterCount> &AllowedRegisters,
-                     const model::Register::Values PointerToCopyLocation,
-                     model::Binary &TheBinary) {
-    if (UsedRegisters.size() == 0) {
-      auto Void = TheBinary.getPrimitiveType(model::PrimitiveTypeKind::Void, 0);
-      return model::QualifiedType{ Void, {} };
+  tryConvertingReturnValue(const SortedVector<RegisterType> &Used,
+                           const RegisterArray<RegisterCount> &Allowed,
+                           const model::Register::Values PointerToCopyLocation,
+                           const model::Binary &Binary,
+                           TypeVector &TypeDependencies) {
+    if (Used.size() == 0) {
+      // Short-circuit the functions that don't return anything (`void`).
+      return ensurePrimitiveTypeIsAvailable(model::PrimitiveTypeKind::Void,
```

Comment:
```suggestion
      return getPrimitiveTypeIfAvailable(model::PrimitiveTypeKind::Void,
```
or
```suggestion
      return tryGetPrimitiveType(model::PrimitiveTypeKind::Void,
```

Reply:
I'm getting rid of it in favour of an asserting `selectPrimitiveType` after our discussion last week. As such I'll close this thread.

Reply:
Nice. I'd just go with `getPrimitiveType`, since AFAIU it does the same thing as the method in Binary.

Reply:
This is all gone. Now type manager takes care of selecting types and making sure the are available.

Closing this thread.